### PR TITLE
Connect resharding finish operation

### PR DESCRIPTION
**Pull request by timvisee**
Tracked in: #4213 
Depends on: <https://github.com/qdrant/qdrant/pull/4507>

When the resharding driver is done, properly end resharding and clean things up.

This is one of the necessary changes we need to complete a resharding operation front to end. I'd like this so we can start testing with it soon.

### Tasks
- [x] Merge <https://github.com/qdrant/qdrant/pull/4507>
- [x] Rebase onto `dev`
- [x] Undraft PR

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### CODE REVIEW

File: lib/collection/src/shards/shard_holder.rs [hunk 1]
```diff
@@ -265,7 +265,14 @@ impl ShardHolder {
             Ok(())
         })?;
 
-        todo!()
+        // TODO(resharding): don't commit resharding here, commit through consensus in driver
+        self.commit_write_hashring(resharding_key.clone())?;
```
Comment (timvisee) [inline lib/collection/src/shards/shard_holder.rs:269]: This is here until we propose hash ring changes through consensus in the driver. That'll be in a separate PR.

**Review by ffuugoo — APPROVED** [review]
I don't like it, but OK. 

Though, I'd structure `finish_resharding` like this:

```rust
    pub fn finish_resharding(&mut self, resharding_key: ReshardKey) -> CollectionResult<()> {
        // TODO(resharding): don't commit resharding here, commit through consensus in driver
        self.commit_write_hashring(resharding_key.clone())?;

        self.check_resharding(&resharding_key, |_| {
            // TODO(resharding): Check resharding is in the correct state to finish resharding!
            Ok(())
        })?;

        self.resharding_state.write(|state| {
            state.take();
        })?;

        Ok(())
    }
```

I.e., put `commit_write_hashring` call *before* `check_resharding`.

**Comment by timvisee** [comment]
> put `commit_write_hashring` call _before_ `check_resharding`.

It panics

**Comment by timvisee** [comment]
> put `commit_write_hashring` call _before_ `check_resharding`.

Then the check panics.

We can also implement the consensus calls for setting resharding states first, because then we don't have to do this.

**Comment by timvisee** [comment]
I moved this on top of <https://github.com/qdrant/qdrant/pull/4507> so that we don't need the temporary 'commit write' hack anymore.

This now just connects the finish resharding consensus message.
